### DISCUSSION TITLE
Extensibility of All Products and AbstractProductGrid product list (e.g. Newest Products) vs Product Collection - next steps

### DISCUSSION THREAD

**Discussion by kmanijak** [woocommerce/woocommerce#48814] [WooCommerce, Blocks, and Site Editing]
# Background

In https://github.com/woocommerce/woocommerce/discussions/44776 we discussed the plans to make the Product Collection block (PC) a main product list block. We also announced it in [Announcing the Product Collection Block](https://developer.woocommerce.com/2024/02/22/announcing-the-product-collection-block/) and we’re making our way toward that goal.

Making Product Collection a main block means we want to hide all the other product list blocks from the inserter. That includes:
- Products (Beta) (https://github.com/woocommerce/woocommerce/pull/48204) - will be included in WooCommerce 9.1.
- All Products
- Blocks based on AbstractProductGrid:
    - Best Selling Products
    - Top Rated Products
    - Newest Products
    - On Sale Products
    - Hand-picked Products
    - Products by Category
    - Products by Tag
    - Products by Attribute

We progressed with Products (Beta) block as Products (Beta) and Product Collection blocks have the same extensibility points. It’s not the case for the other blocks.

# Goal

The goal of the discussion is to gather existing extensibility points of the remaining blocks, provide recommendations on how to approach them in the PC block, and discuss the next steps in the context of soft depreciation (removing from the inserter).

# Extensibility points

Below, I gathered extensibility points. The recommended next steps are color-coded:
- 🟢 - I recommend we port it to Product Collection,
- 🟠 - I recommend we don't port it to Product Collection but there's an alternative approach,
- 🔴 - I recommend we don't port it and there's no alternative.

## All Products

Three experimental actions are coming from the All Products block OR its inner blocks:

### 🟢 JS action `experimental__woocommerce_blocks-cart-add-item`

- **Source:** [Add to Cart button](https://github.com/woocommerce/woocommerce/blob/408af71b279549f07efd314c9ed02ef32ae7adba/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/block.tsx#L75).

- **Description:** Fired when product is added to the cart. Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

- **Usage:**
    - Google Analytics for WooCommerce
    -  [WP Directory search](https://wpdirectory.net/search/01J0T4SSVX6TC4GFJDXAT9WZE6) shows 5 extensions:
        - Google Listings & Ads
        - WooPayments: Integrated WooCommerce Payments
        - Menu Cart for WooCommerce
        - Pixel Manager for WooCommerce – Track Google Analytics, Google Ads, TikTok and more
        - WP Menu Cart

- **Product Collection context:** PC uses the Add to Cart Button in the SSR version (PHP). Block is powered by Interactivity API so the JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.

- **Recommended next step:** implement in PC block.

---

### 🟢 / 🔴 JS action `experimental__woocommerce_blocks-product-view-link`

- **Source:**
    - [Add to Cart Button](https://github.com/woocommerce/woocommerce/blob/408af71b279549f07efd314c9ed02ef32ae7adba/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/block.tsx#L68)
    - [Product Image](https://github.com/woocommerce/woocommerce/blob/0a411dd352f8c8f4bb775cef5b76124606792f35/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/image/block.tsx#L145) 
    - [Product Title](https://github.com/woocommerce/woocommerce/blob/0a411dd352f8c8f4bb775cef5b76124606792f35/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/title/block.tsx#L104)

- **Description:** Fired when product link is clicked (it can be "View More", or "Select Options" buttons, Product Title, or Product Image). Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

- **Usage:**
    - Google Analytics for WooCommerce
    -  [WP Directory search](https://wpdirectory.net/search/01J0RBAXJWF7PVE7DVCYJ1G4ED) shows 1 extension:
        - Pixel Manager for WooCommerce – Track Google Analytics, Google Ads, TikTok and more

- **Product Collection context:**
    - Add to Cart Button - PC uses the Add to Cart Button in the SSR version (PHP). Block is powered by Interactivity API so the JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.
    - Product Image - PC uses the Product Image in the SSR version (PHP). The JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.
    - Product Title - Product Collection doesn’t use the Product Title block but the `core/post-title` variation instead. I don't think we can extend the on-click callback in the Core block. Also, I didn't investigate but for some reason, this event doesn’t even fire for All Products.

- **Recommended next step:**
    - 🟢 Add to Cart Button - implement in PC block.
    - 🟢 Product Image - implement in PC block.
    - 🔴 Product Title - I don't see a good way to port this event other than migrating Core's Post Title block to Interactivity API or using WooCommerce's Product Title block instead. Hence due to technical limitations, I suggest we skip this one **for now**. It's not an excuse or good argument but it's worth mentioning that it's not fired from any other product list blocks at the moment.
---

### 🟢 JS action `experimental__woocommerce_blocks-product-list-render`

- **Source:** [All Products](https://github.com/woocommerce/woocommerce/blob/29d4b18a8546aa8b44a55fe6f0de980849774178/plugins/woocommerce-blocks/assets/js/base/components/product-list/product-list.tsx#L166)

- **Description:** Fired when All Products block is rendered or re-rendered. Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

- **Usage:**
    - Google Analytics for WooCommerce
    -  [WP Directory search](https://wpdirectory.net/search/01J0RBBMAM04ET60E00KK7HHSB) shows 2 extensions:
        - Pixel Manager for WooCommerce – Track Google Analytics, Google Ads, TikTok and more
        - Analytics & Privacy Toolkit – WP Full Picture

- **Product Collection context:** PC doesn’t fire this action. We could do it in the same manner as `AbstractProductGrid` does (check the next section) although it attaches DOM events listeners, which we should be avoiding with the Interactivity API in mind. Worth checking if we could utilize Interactivity API for that (e.g. [`wp-on-document-async`](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/api-reference/#wp-on-document-async)).

- **Recommended next step:** implement in PC block.

There’s an issue https://github.com/woocommerce/woocommerce/issues/42985 I created some time ago. The goal is to make a decision if we want to keep supporting experimental actions or not (including the ones mentioned above). If we want to keep them (which I think so), we should also answer the question of whether to graduate them to stable.
At the same time, there’s an ongoing discussion by @alexflorisca about experimental APIs (https://github.com/woocommerce/woocommerce/discussions/43091).

---

### 🟢 JS action `experimental__woocommerce_blocks-product-list-render`

- **Source:** [AbstractProductGrid](https://github.com/woocommerce/woocommerce/blob/bb9b332ae2003310549a24f221f345be78f2de7a/plugins/woocommerce/src/Blocks/BlockTypes/AbstractProductGrid.php#L94-L126)

- **Description:** Fired when AbstractProductGrid is rendered (once per single page load). Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

- **Usage:** The same case as in All Products ⏫ 

- **Product Collection context:** The same case as in All Products ⏫ 

- **Recommended next step:** The same case as in All Products ⏫ 

---

### 🔴 PHP filter `woocommerce_blocks_product_grid_add_to_cart_attributes`

- **Source:** [AbstractProductGrid](https://github.com/woocommerce/woocommerce/blob/29d4b18a8546aa8b44a55fe6f0de980849774178/plugins/woocommerce/src/Blocks/BlockTypes/AbstractProductGrid.php#L679)

- **Description:** Filter to modify the HTML attributes of add to cart block which is an anchor element (`<a/>`) in `AbstractProductGrid`.

- **Usage:** [WP Directory search](https://wpdirectory.net/search/01J0RB6S0D6EBNGQQ0R09RCW5J) shows 1 extension:
    - Gift Vouchers for WooCommerce

- **Product Collection context:** Add to Cart Button doesn’t expose the filter to HTML element attributes (`target` etc.). But also, there’s no need for that. The experience is curated in PC block, with no redirect/reload and powered by Interactivity API.

- **Recommended next step:** Do not be implement in PC block as it doesn't match the current setup.

---

### 🔴 PHP filter `woocommerce_blocks_product_grid_is_cacheable`

- **Source:** [AbstractProductGrid](https://github.com/woocommerce/woocommerce/blob/29d4b18a8546aa8b44a55fe6f0de980849774178/plugins/woocommerce/src/Blocks/BlockTypes/AbstractProductGrid.php#L354)

- **Description:** Determines if `AbstractProductGrid` is cacheable which results in getting posts either through `$query->get_cached_posts` (cacheable) or `$query->get_posts` (not cacheable).

- **Usage:** [WP Directory search](https://wpdirectory.net/search/01J0RB4RB12DRQHV0W5PC4FXZN) shows 4 extensions:
    - YITH WooCommerce Ajax Product Filter
    - Product Filter by WBW
    - MultiVendorX Marketplace – WooCommerce MultiVendor Marketplace Solution
    - Filkers – Turn your products into videos

- **Product Collection context:** The filter is specific to `AbstractProductGrid` implementation which uses `get_cached_posts` while Product Collection doesn’t. It may be a optimization to consider in PC block but until we use `get_cached_posts` there’s no space for that filter.

- **Recommended next step:** Do not be implement in PC block as it's not valid in context of this block.

---

### 🟠 PHP filter `woocommerce_blocks_product_grid_item_html`

- **Source:** [AbstractProductGrid](https://github.com/woocommerce/woocommerce/blob/29d4b18a8546aa8b44a55fe6f0de980849774178/plugins/woocommerce/src/Blocks/BlockTypes/AbstractProductGrid.php#L513)

- **Description:** Filter to modify the HTML output. `AbstractProductGrid` blocks didn’t offer Editor experience to modify the content except for a couple of settings toggling on/off elements like image or title. So filter offered a way to introduce some changes in layout, functionality, styling.

- **Usage:** [WP Directory search](https://wpdirectory.net/search/01J0RB863JZFNE8EVXYNFSQN43) shows 44 extensions.

- **Product Collection context:** PC doesn’t expose a dedicated filter allowing to amend a product HTML. But at the same time, it’s not needed as Editor gives huge flexibility to configure Product Template which wasn’t the case with `AbstractProductGrid`.

- **Recommended next step:** Considering the paradigm shift towards blocks and rich flexibility of PC block in Editor we don’t plan to polyfill/support a dedicated HTML filter like this one. But considering the high usage of the filter we provide our recommendations on achieving the same results with the PC block.

The order of recommendation below matters, move to next proposition only if the previous doesn’t satisfy your needs.

---

##### 1. Use Editor and WooCommerce blocks

`AbstractProductGrid` blocks provided curated layout with few toggles to show/hide some elements. PC block provides much richer editing experience: there’s more product elements to choose from (e.g. Product Stock Indicator, Product SKU etc.), elements can be rearranged and each of them provide a set of styling options.

**Use Product Collection Editor experience in the following use cases:**
  - Changing the product layout (adding/removing elements or changing order),
  - Changing the styling (font sizes, colors, etc.),
  - Adding functionality covered by existing blocks (e.g. Product Stock Indicator). Remember that not all product elements are included in the PC block by default. You can check the whole list by typing “Product Elements” in the inserter when focused on Product Template block

<details>
  <summary>Example: Displaying the product stock information</summary>

#### Before

Use `woocommerce_blocks_product_grid_item_html` and modify the HTML directly to add the element responsible for displaying the stock information. For example:

```
    function display_stock_status($html, $data, $product) {
            $stock_status = $product->get_stock_status();
            $new_html       = $html . ‘<br /><p>' . $stock_status . ‘</p>’;

            return $new_html;
    }

    add_filter('woocommerce_blocks_product_grid_item_html', ‘display_stock_status’), 10, 3 );
```

#### With PC block:

Go to Editor and include Product Stock Indicator block in the Product Template

</details>

---

##### 2. Build custom blocks

If there’s functionality not covered by WooCommerce blocks you can develop a custom block (follow the [tutorial](https://developer.wordpress.org/block-editor/getting-started/tutorial/)) and use it in PC block as an inner block.

**Develop custom block in the following use cases:**
  - Gutenberg and WooCommerce (or other plugins you use) don’t provide the functionality you need

<details>
  <summary>Example: Displaying the stock quantity of products</summary>

#### Before

Use `woocommerce_blocks_product_grid_item_html` and modify the HTML directly to add the element responsible for displaying the stock information. For example:

```
    function display_stock_quantity($html, $data, $product) {
            // The function can be used in your custom block
            $stock_quantity= $my_custom_function_creating_stock_quantity_content();
            $new_html         = $html . ‘<br />' . $stock_quantity;

            return $new_html;
    }

    add_filter('woocommerce_blocks_product_grid_item_html', ‘display_stock_quantity’), 10, 3 );
```

#### With PC block:

After building a new block, go to Editor and use it in the Product Template. Take advantage of the Editor and arrange the layout and use out-of-the-box global styles.

</details>

---

##### 3. Use Block Hooks API

Previous options require manually inserting blocks in the Editor. This can be automated thanks to [Block Hooks API](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/). It allows a block to insert itself relative to instances of other block types automatically. One of the biggest advantages is you can see the outcome not only in frontend as it was with the previous filter, but also in Editor.

**Use Block Hooks API in the following use cases:**
- You want to automatically insert some block into Product Collection

<details>
  <summary>Example: Your extension displays “Like” button next to price</summary>

#### Before

Use `woocommerce_blocks_product_grid_item_html` and modify the HTML directly. it requires finding the element you need (error prone) and manual attachment of block. For example:

```
    function add_like_button($html, $data, $product) {
            $like_button = $get_my_like_button();
            $price_div    = '<div class="wp-block-button wc-block-grid__product-price">';
            $new_html    = preg_replace('/'.$price_div.'(.*)<\/div>/', $price_div.'</div>'.$like_button, $html);

            return $new_html;
    }

    add_filter('woocommerce_blocks_product_grid_item_html', ‘add_like_button’), 10, 3 );
```

#### With PC block:

When having “Like” button block, follow [Block Hooks API example](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/) to inject it somewhere in the Product Template block:

```
function block_hooks_like_button($hooked_blocks, $position, $anchor_block, $context) {

	// Extend only Product Collections in Templates (e.g. Product Catalog)
	if ($context instanceof WP_Block_Template) {

		// Hooks the "Like" button block before the Product Price block
                //  in the Product Catalog template.
		if (
			'woocommerce/product-price' === $anchor_block &&
			'before' === $position &&
			'archive-product' === $context->slug
		) {
			$hooked_blocks[] = 'ockham/like-button';
		}
	}

	return $hooked_blocks;
}

add_filter('hooked_block_types', ‘block_hooks_like_button’, 10, 4);
```

Here's an example outcome:

Editor | Frontend
-- | --
<img width="830" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/5d83e0ca-8c25-4a81-b1e3-d29e4918f28f"> | <img width="929" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/e627242e-6e72-4cde-8d12-1d9fac929b58">

</details>

---

##### 4. Use `render_block_*` filter

If your use case really cannot be satisfied by all the previous options, you may consider using `render_block_woocommerce/product-collection` or `render_block_woocommerce/product-template` filters. We suggest it as a last option.

**Use render_block filters in the following use cases:**
- The change is not visual and cannot cover by blocks
- The change requires interference in blocks/HTML structure

<details>
  <summary>Example: Disable client-side navigation powered by Interactivity API</summary>

#### Before

Use `woocommerce_blocks_product_grid_item_html` and modify the HTML directly.

#### With PC block:

Use `render_block_woocommerce/product-collection` to detect 3rd party blocks and modify the HTML output. It's the real case from Product Collection [here](https://github.com/woocommerce/woocommerce/blob/841d00e1506f43dd31c1f66078b93f1edea5aaa0/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L80).

When using these filters we recommend using [`WP_HTML_Tag_Processor`](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/) for HTML modifications.

</details>

---

5. Upcoming: register custom collections

We’re working on the API (`__experimentalRegisterProductCollection`) that will allow registering custom collections. You’ll have the ability to register and expose to users a preconfigured PC block. Configuration will include:
- Layout and inner blocks structure - it may by default include your custom blocks,
- `query` attributes - you define the criteria what products to display,
- and more.
Stay tuned!

---

# Summary

We can see there are two types of extensibility points:

1. JS actions with the primary purpose of analytics but also, there were voices in WooCommerce Community Slack that the events are missing from Product Collection block.

**Recommendation:**
🟢 Graduate the events to stable (they haven’t changed for a long time) and implement them (2 out of 3) in Product Collection before soft-deprecating All Products and AbstractProductGrid blocks.
Alternatively, at least we should implement them in PC without graduation if we don’t do it in the scope of PC development.

2. PHP filters with the primary purpose of editing AbstractProductGrid blocks content. They were exposed when the Editor capabilities and flexibility were limited. Product Collection is super-customisable in comparison to AbstractProductGrid hence there’s no need for them.

**Recommendation:**
- 🔴 `woocommerce_blocks_product_grid_add_to_cart_attributes` - drop the support - there’s low usage of the filter and the use case doesn’t match the current setup anymore.
- 🔴  `woocommerce_blocks_product_grid_is_cacheable` - drop the support - it’s no longer valid in the context of PC block.
- 🟠 `woocommerce_blocks_product_grid_item_html` - don’t expose such a filter in PC block. Depending on the use case, users should try using existing blocks and leverage Editor experience, build custom blocks, or use Block Hooks API. If really needed, `render_block_woocommerce/product-collection` and `render_block_woocommerce/product-template` can be used. In future, the extensibility will be even simpler with possibility to register custom collections.

> [!NOTE]  
> All Products: only the JS actions are required to allow soft-deprecating.
> AbstractProductGrid: both JS actions and PHP filters are required to allow soft-deprecating.

# Feedback needed

I’m seeking feedback on the above:
- Are there any other extensibility points that are in use that I missed?
- What do you think about recommended actions?

---

cc: @imanish003 , @ObliviousHarmony , @xristos3490 , @sunyatasattva , @danielwrobert

**Comment by kmanijak** [comment]
Here are the issues to cover JS actions in PC block:

- `experimental__woocommerce_blocks-cart-add-item` - https://github.com/woocommerce/woocommerce/issues/42924
- `experimental__woocommerce_blocks-product-view-link` - https://github.com/woocommerce/woocommerce/issues/48861
- `experimental__woocommerce_blocks-product-list-render` - https://github.com/woocommerce/woocommerce/issues/48862

**Comment by imanish003** [comment]
Hey Karol!

Great write-up! I really appreciate the effort you put into this. It's very clear and easy to understand 🚀 I have a few comments on this:

> 🔴 Product Title - I don't see a good way to port this event other than migrating Core's Post Title block to Interactivity API or using WooCommerce's Product Title block instead. Hence due to technical limitations, I suggest we skip this one for now. It's not an excuse or good argument but it's worth mentioning that it's not fired from any other product list blocks at the moment.

IMO Either we should track it for all the cases or skip it for all the cases for now. As the name `experimental__woocommerce_blocks-product-view-link` suggest, we are tracking if the product is viewed or not. It makes sense to track it if product is viewed by clicking on Product Title.
- Can we skip this action for all the cases for now?
  - If yes, then we can create a new issue for tracking and handle it later. 
  - If no, then maybe we should consider adding this tracking to the Product Collection block. As per my understanding, it will be a high-effort task, so we should think about it before making any decision.

But it's just my opinion. Hopefully, others input will make it more clear on what we should do with this.

> The goal is to make a decision if we want to keep supporting experimental actions or not (including the ones mentioned above). If we want to keep them (which I think so)

I am with you on this as well. We definitely want to keep actions which helps with analytics. There are few questions we need to ask ourselves:
- Are these actions helpful for analytics? IMO Yes, We definitely want to track if the product is viewed or added to cart. But I am not sure about `experimental__woocommerce_blocks-product-list-render`. Do we really need to track when the product list is rendered? In future, we can track which collections are rendered most and which are not. Which would be more helpful for us to decide which collections are more popular and which are not. 
- Are these actions helpful for extension developers? IMO Yes, That's why these are being used by multiple extensions.

> 🔴 PHP filter woocommerce_blocks_product_grid_add_to_cart_attributes

FWIW, I believe this can be achieved using "Use render_block_* filter," which you explained in the Github Discussion. 

> Product Collection context: The filter is specific to AbstractProductGrid implementation which uses get_cached_posts while Product Collection doesn’t. It may be a optimization to consider in PC block but until we use get_cached_posts there’s no space for that filter.

Makes sense to me. However, It would be interesting to explore how we can use get_cached_posts in the Product Collection block. And how much performance improvement we can get by using it. Looks like it was implemented by @mikejolley in [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/1086/). It's just a [simple wrapper](https://github.com/woocommerce/woocommerce/blob/f48439e02d36893d47cb74aec7f0ba02321d55e4/plugins/woocommerce/src/Blocks/Utils/BlocksWpQuery.php#L14) over WP_Query. So, I think it should be easy to implement in the Product Collection block as well.

> Product Collection equivalents to woocommerce_blocks_product_grid_item_html

Wow! That's a great explanation. I really like the sequence of recommendations. It's very clear and easy to understand. It will be helpful for developers who are new to blocks. Maybe we should add this to our developer documentation as well.

**Reply by kmanijak** [reply]
Thanks for reply!

> IMO Either we should track it for all the cases or skip it for all the cases for now. As the name experimental__woocommerce_blocks-product-view-link suggest, we are tracking if the product is viewed or not. It makes sense to track it if product is viewed by clicking on Product Title.

@gigitux suggested we could use `render_block` to detect if the Product Title is in Product Collection and inject Interactivity API to trigger an event. So basically let's go for "we should track it for all the cases".

> I am not sure about experimental__woocommerce_blocks-product-list-render [...] Are these actions helpful for extension developers? IMO Yes, That's why these are being used by multiple extensions.

It's fairly easy to implement so I suggest we implement that for compatibility reasons.

**Comment by nerrad** [comment]
One quick thought I had when looking through the JS actions. Have we considered exposing custom native JS events that extensions can hook into?

**Reply by kmanijak** [reply]
Thanks for your reply! I didn't consider that when preparing this discussion but at the same time, I didn't see a need for that. Is there a reason we should consider moving away from `@wordpress/hooks`'s `doAction` method?

**Comment by gigitux** [comment]
Thanks, @kmanijak, for kicking off this discussion!

> JS action experimental__woocommerce_blocks-cart-add-item
>  PC uses the Add to Cart Button in the SSR version (PHP). Block is powered by Interactivity API so the JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.

Regarding this action, I think that it is not necessary to port it to the Product Collection. This event is triggered for the JS version of the Product Button that is run only on the All Products block. 

The Product Button used in the Product Collection is powered by the Interacitity API. In the [`addToCart` function](https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx/#L135-L148), `addItemToCart` is invoked that trigger two custom native JS events:
- `wc-blocks_adding_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L281-L282
- `wc-blocks_added_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L292-L293

> IMO Either we should track it for all the cases or skip it for all the cases for now. As the name experimental__woocommerce_blocks-product-view-link suggest, we are tracking if the product is viewed or not. It makes sense to track it if product is viewed by clicking on Product Title.

I agree with @imanish003. This tracking could be beneficial. We could try using the `render_block` filter and inject the interactivity API only when the Product Title is an inner block of the Product Collection block.

In my opinion, we should implement all necessary filters/events for tracking but avoid implementing any filter that allows customization of the DOM structure. All the blocks already have the render_block filter for updating the HTML structure. 


>  PHP filter woocommerce_blocks_product_grid_item_html

PS: It's so cool to see how much the product collection is extensible! 😍 Huge kudos to all the folks involved!

**Reply by kmanijak** [reply]
Thanks for your reply Luigi!

> The Product Button used in the Product Collection is powered by the Interacitity API. In the [addToCart function](https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx/#L135-L148), addItemToCart is invoked that trigger two custom native JS events:
> - wc-blocks_adding_to_cart
> - wc-blocks_added_to_cart

I wasn't aware of that! That is a nice alternative that should be promoted instead!

> I agree with @imanish003. This tracking could be beneficial. We could try using the render_block filter and inject the interactivity API only when the Product Title is an inner block of the Product Collection block.

Nice! Thanks for your input. Let's do that.

> In my opinion, we should implement all necessary filters/events for tracking but avoid implementing any filter that allows customization of the DOM structure. All the blocks already have the render_block filter for updating the HTML structure.

Couldn't agree more!

**Comment by kmanijak** [comment]
Thanks all for your participation, @gigitux , @imanish003 , @nerrad . I want to sum up and close the discussion.

Let's cover all three events that are triggered from various Product Elements:

- experimental__woocommerce_blocks-cart-add-item - https://github.com/woocommerce/woocommerce/issues/42924
- experimental__woocommerce_blocks-product-view-link - https://github.com/woocommerce/woocommerce/issues/48861
- experimental__woocommerce_blocks-product-list-render - https://github.com/woocommerce/woocommerce/issues/48862

@gigitux noted there's an alternative to `experimental__woocommerce_blocks-cart-add-item`, however, after giving it some thoughts I think we should keep the events symmetrical. So, either support all of them or not. I don't want 3PD needed to use different stuff for similar things, e.g.:
- `wc-blocks_added_to_cart` - JS event
- `experimental__woocommerce_blocks-product-view-link` - action

**Reply by gigitux** [reply]
I don't have a strong opinion on this, so I'm fine with your final conclusion. However, I’d like to share my final thoughts.

I believe that maintaining two methods to track the same event increases cognitive load:

- For third-party developers (3PDs), as they might question why there are two ways to track the same events.
- For developers working on this codebase, as it can cause confusion and redundancy.


Futhermore, these actions are "experimental" and given that the Product Collection is a stable block, we should upgrade them IMHO. 

I'm more inclined to remove the action, tbh!

**Reply by kmanijak** [reply]
After thinking it through I agree with you and consolidating yours and @nerrad's feedback I think we should:
- keep and promote: `wc-blocks_added_to_cart` event instead of `experimental__woocommerce_blocks-cart-add-item`
- create similar events as alternatives to experimental actions:
    - `wc-blocks_viewed_product` instead of `experimental__woocommerce_blocks-product-view-link`
    -  `wc-blocks_product_list_rendered` instead of `experimental__woocommerce_blocks-product-list-render`.

Does it make sense? Happy to hear some other opinions: @imanish003 , @ObliviousHarmony , @WunderBart.

BTW I reopened the discussion to finalize that new thread.

**Reply by imanish003** [reply]
Makes sense to me. 

My only concern is, are we okay with not providing same extensibility points as for older blocks? Because if we change our approach from using JS actions to JS events, we are not providing the same extensibility points as before. Plugins will need to make changes to their code to track the same events.
- Plugins will need to use JS actions for older blocks
  - OR we are planning to replace JS actions with JS events for older blocks as well?
- Plugins will need to use JS events for new version of WooCommerce

As @gigitux mentioned, older events are experimental so we can actually make changes to them. if that's okay, then sure, we can go ahead with this approach of using JS events. 

Also, I believe we need to understand why we want to move from JS actions to JS events. Is there any specific reason for that? 
- Is it to keep ourselves consistent with events like `wc-blocks_added_to_cart`?
- OR, are there actually some benefits of using JS events over JS actions?

**Reply by kmanijak** [reply]
After having some discussions elsewhere the consensus is:

- promote: `wc-blocks_added_to_cart `event instead of `experimental__woocommerce_blocks-cart-add-item`
- create similar events as alternatives to experimental actions:
    - `wc-blocks_viewed_product` instead of `experimental__woocommerce_blocks-product-view-link`
    - `wc-blocks_product_list_rendered` instead of `experimental__woocommerce_blocks-product-list-render`.

Product Collection doesn't have to be backward compatible with older blocks but should provide a feature parity, hence it's ok if we provide alternative ways to react to the same events. And JS events are more reliable than JS filters.

That being said,
- I'll close https://github.com/woocommerce/woocommerce/issues/42924
- and repurpose https://github.com/woocommerce/woocommerce/issues/48862 and https://github.com/woocommerce/woocommerce/issues/48861

In addition, we need to publish a dev advisory for that.

I'd love to get some final confirmation you agree with that @gigitux , @imanish003 , @xristos3490 , @ObliviousHarmony before proceeding and closing this discussion again 🙌

**Reply by imanish003** [reply]
@kmanijak Sounds good to me! I agree with the proposed changes. Thanks for moving this forward :rocket:

**Reply by kmanijak** [reply]
Update:
- Issue closed as not planned:
    - https://github.com/woocommerce/woocommerce/issues/42924
- Issues repurposed as per discussion:
    - https://github.com/woocommerce/woocommerce/issues/48862
    - https://github.com/woocommerce/woocommerce/issues/48861


I'm closing the discussion again, thanks all for your participation! 🙌

**Reply by medavidallsop** [reply]
Just looking through this now as saw the dev advisory blog post yesterday.

So lets say currently you are using `woocommerce_after_shop_loop_item` to insert some text after the shop loop item conditionally based on some PHP conditions of the `$product`

- Old product blocks (other than All Products), you could use woocommerce_blocks_product_grid_item_html to effectively achieve the same as what you're doing on `woocommerce_after_shop_loop_item` albeit with some duplicated code
- Product Collection block on template based pages will work due to the compatibility layer
- Product Collection block on standalone pages, the compatibility layer will not work, and therefore it appears you have to use `render_block_woocommerce/product-collection` to somehow iterate through each product and attempt to add it in that way

However, for the standalone pages scenario, `render_block_woocommerce/product-collection` is also going to happen on the template based pages where the compatibility layer is working, is there a condition which can be used in `render_block_woocommerce/product-collection` to only apply the modifications if it's not a block template or similar to stop the duplication...?

Also if anyone has any examples of using `render_block_woocommerce/product-collection` to add some content after each product, that would be great - is there maybe something more specific to target each product within the product collection?

FYI: I've just created a discussion on parts of this here: https://github.com/woocommerce/woocommerce/discussions/52915

**Reply by kmanijak** [reply]
Hey @medavidallsop, thanks for your response!

> However, for the standalone pages scenario, `render_block_woocommerce/product-collection` is also going to happen on the template based pages where the compatibility layer is working

I'm thinking now, if you want to achieve the same result in templates and posts/pages you could ONLY rely on `render_block_woocommerce/product-collection` and ignore compatibility layer completely. That means you'd target ALL Product Collection blocks with single implementation. Is there anything else that stops you from this? 🤔 

> is there a condition which can be used in `render_block_woocommerce/product-collection` to only apply the modifications if it's not a block template or similar to stop the duplication...?

If you cannot rely only on `render_block_woocommerce/product-collection` you can use WordPress helpers like:

- [`is_archive()`](https://developer.wordpress.org/reference/functions/is_archive/): True on any archive page
- [`is_singular()`](https://developer.wordpress.org/reference/functions/is_singular/): True on any single post/page

**Reply by medavidallsop** [reply]
>I'm thinking now, if you want to achieve the same result in templates and posts/pages you could ONLY rely on render_block_woocommerce/product-collection and ignore compatibility layer completely. That means you'd target ALL Product Collection blocks with single implementation. Is there anything else that stops you from this? 🤔

Yes, as the compatibility layer must be used. There isn't really a possibility of ignoring the compatibility layer (not sure how you would as it is on by default, and you wouldn't wanna filter it to be disabled as other plugins may rely on it), it must be used because it ensures any functionality/content that was previous added via the classic hooks work as expected - for websites using classic themes and shortcodes, and websites using block themes and product collection blocks, and ensures a developer doesn't really have to do anything to make their functionality/content from classic hooks compatible with the product collection block.

IMO the best solution here is for the compatibility layer to work in a standalone context, as without it the product collection block doesn't allow for the same extensibility as `[products]` shortcode provides. I've documented this more here: https://github.com/woocommerce/woocommerce/discussions/52915
